// lavaclient/dist/lib/node/REST.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.REST = void 0;
const undici_1 = require("undici");
class REST {
    constructor(node) {
        this.node = node;
        this.pool = new undici_1.Pool(this.baseUrl);
    }
    get info() {
        return this.node.conn.info;
    }
    get baseUrl() {
        return `http${this.info.secure ? "s" : ""}://${this.info.host}:${this.info.port}`;
    }
    loadTracks(identifier) {
        return this.do(`/loadtracks?identifier=${encodeURIComponent(identifier)}`);
    }
    decodeTracks(...tracks) {
        return this.do("/decodetracks", { method: "POST", data: JSON.stringify(tracks) });
    }
    decodeTrack(track) {
        return this.do(`/decodetrack?track=${track}`);
    }
    async do(endpoint, { method = "GET", data } = {}) {
        endpoint = /^\/.+/.test(endpoint) ? endpoint : `/${endpoint}`;
        try {
            const request = await this.pool.request({
                path: endpoint,
                method: method,
                body: data ? JSON.stringify(data) : undefined,
                headers: {
                    "Authorization": this.info.password,
                    "Client-Name": this.node.conn.clientName,
                },
            });
            this.node.debug("rest", `+ ${method} ${endpoint}`);
            return await request.body.json();
        }
        catch (e) {
            this.node.emit("error", e instanceof Error ? e : new Error(`${e}`));
            this.node.debug("rest", `- ${method} ${endpoint}`);
            throw e;
        }
    }
}
exports.REST = REST;
