// lavaclient/dist/lib/track/DataInput.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataInput = void 0;
const node_util_1 = require("node:util");
const decoder = new node_util_1.TextDecoder();
class DataInput {
    constructor(data) {
        this.pos = 0;
        if (typeof data === "string")
            data = new Uint8Array(Buffer.from(data, "base64"));
        this.buf = data;
        this.view = new DataView(this.buf.buffer);
    }
    readByte() {
        return this.view.getUint8(this.advance());
    }
    readBoolean() {
        return this.readByte() !== 0;
    }
    readUnsignedShort() {
        return this.view.getUint16(this.advance(2));
    }
    readInt() {
        return this.view.getInt32(this.advance(4));
    }
    readLong() {
        return this.view.getBigInt64(this.advance(8));
    }
    readUTF() {
        const length = this.readUnsignedShort(), start = this.advance(length);
        return decoder.decode(this.buf.slice(start, start + length));
    }
    advance(by = 1) {
        if (this.pos + by > this.buf.length) {
            throw new Error(`EOF: unable to read ${by} bytes.`);
        }
        const cpos = this.pos;
        this.pos += by;
        return cpos;
    }
}
exports.DataInput = DataInput;
