// lavaclient/dist/lib/track/Track.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.decode = exports.decoders = exports.TRACK_INFO_VERSION = exports.TRACK_INFO_VERSIONED = void 0;
const DataInput_1 = require("./DataInput");
exports.TRACK_INFO_VERSIONED = 1, exports.TRACK_INFO_VERSION = 2;
exports.decoders = {
    2: input => {
        const track = {
            title: input.readUTF(),
            author: input.readUTF(),
            length: Number(input.readLong()),
            identifier: input.readUTF(),
            isStream: input.readBoolean(),
            uri: input.readBoolean() ? input.readUTF() : "",
            sourceName: input.readUTF(),
            version: 2,
        };
        if (["local", "http"].includes(track.sourceName ?? "")) {
            const [name, parameters] = input.readUTF().split("|");
            track.probeInfo = { name, parameters };
        }
        track.position = Number(input.readLong());
        track.isSeekable = !track.isStream;
        return track;
    },
};
function readVersion(input) {
    const flags = (input.readInt() & 0xC0000000) >> 30;
    return flags & exports.TRACK_INFO_VERSIONED ? input.readByte() : 1;
}
function decode(data) {
    const input = new DataInput_1.DataInput(data), version = readVersion(input);
    return exports.decoders[version]?.(input) ?? null;
}
exports.decode = decode;
